pywbem/cim_obj.py

```diff
@@ -2511,6 +2511,72 @@ def to_wbem_uri(self, format='standard'):
 
         return _ensure_unicode(''.join(ret))
 
+    @staticmethod
+    def from_instance(class_, instance, namespace=None, host=None,
+                      strict=False):
+        """
+        Given a class and corresponding instance, create and return the
+        CIMInstanceName from the class key properties and instance key
+        property values.
+
+        If a property value does not exist in the instance it gets the
+        value from the class default_value unless strict is set. If strict
+        is True, all key properties must exist in the instance
+
+        Parameters:
+
+            class_ (:class:~pywbem.CIMClass`):
+                CIMClass from which the key properties will be retrieved.
+
+            instance: (:class:~pywbem.CIMInstance`):
+                CIMInstance containing the key property values.
+
+            namespace: (:term:``string`):
+                Namespace to include in created CIMInstanceName or None.
+
+            host:  (:term:``string`):
+                Host name to include in created CIMInstanceName or None.
+
+            strict: (:class:`py:bool`):
+                If True, all key properties are required in instance. Default
+                False
+
+        Returns:
+            :class:`CIMInstanceName` built from the key properties in the
+            class using the key property values in the instance.
+
+        Raises:
+            ValueError: Class and instance names do not match or the
+            strict attribute is True and key property does not exist in
+            the instance.
+
+        """
+        keybindings = NocaseDict()
+        if class_.classname != instance.classname:
+            raise ValueError('Class name %s does not match instance '
+                             'classname %s' %
+                             (class_.classname, instance.classname))
+
+        for prop in class_.properties:
+            if 'key' in class_.properties[prop].qualifiers:
+                pname = class_.properties[prop].name  # get original name
+
+                if prop in instance:
+                    keybindings[pname] = instance[prop]
+                else:
+                    if strict:
+                        raise ValueError('Key Property %s in class %s '
+                                         'but not in instance.' %
+                                         (pname, class_.classname))
```

Comment:
I don't like this error message; it seems to suggest that a key property can stand on its own and can be in a class or not. How about: "Key property %r of class %r is missing in the instance.". Always use `%r` instead of `%s` for CIM names, because unicode characters will be properly escaped that way, if needed.

Reply:
Done and note about discussion.

Reply:
**DISCUSSION:** the general solution to logging, etc. of CIMNames.  One question is whether logging is generally protective of issues with unicode in name displays.